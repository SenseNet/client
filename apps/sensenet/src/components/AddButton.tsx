import { Schema } from '@sensenet/default-content-types'
import { useLogger, useRepository } from '@sensenet/hooks-react'
import {
  createStyles,
  IconButton,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  makeStyles,
  Popover,
  Tooltip,
} from '@material-ui/core'
import { CloudUploadOutlined } from '@material-ui/icons'
import Add from '@material-ui/icons/Add'
import clsx from 'clsx'
import React, { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { globals, useGlobalStyles } from '../globalStyles'
import { useLocalization, usePersonalSettings, useQuery, useSnRoute } from '../hooks'
import { navigateToAction } from '../services'
import { useDialog } from './dialogs'
import { Icon } from './Icon'

const useStyles = makeStyles(() => {
  return createStyles({
    addButtonDisabled: {
      backgroundColor: '#CCCCCC !important',
      '& svg': {
        color: '#8C8C8C',
      },
    },
    listItem: {
      width: '100%',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-evenly',
      height: globals.common.drawerItemHeight,
    },
    listDropdown: {
      padding: '10px 0 10px 10px',
      width: '245px',
      maxHeight: '548px',
    },
    listItemTextDropdown: {
      margin: 0,
      textOverflow: 'ellipsis',
      overflow: 'hidden',
      whiteSpace: 'nowrap',
      maxWidth: '139px',
    },
    disabled: {
      cursor: 'not-allowed',
    },
  })
})
export interface AddButtonProps {
  isOpened?: boolean
}

export const AddButton: React.FunctionComponent<AddButtonProps> = (props) => {
  const classes = useStyles()
  const globalClasses = useGlobalStyles()
  const repo = useRepository()
  const history = useHistory()
  const { openDialog } = useDialog()
  const [showSelectType, setShowSelectType] = useState(false)
  const [allowedChildTypes, setAllowedChildTypes] = useState<Schema[]>([])
  const localization = useLocalization().addButton
  const logger = useLogger('AddButton')
  const [isAvailable, setAvailable] = useState(false)
  const [anchorEl, setAnchorEl] = React.useState<HTMLElement | null>(null)
  const [hasUpload, setHasUpload] = useState(false)
  const personalSettings = usePersonalSettings()

  const snRoute = useSnRoute()
  const pathQuery = useQuery().get('path')
  const currentPath = `${snRoute.path ?? ''}${pathQuery ?? ''}`
  const activeAction = snRoute.match?.params.action

  useEffect(() => {
    const getActions = async () => {
      try {
        const actions = await repo.getActions({ idOrPath: currentPath })
        const isActionFound = actions.d.results.some((action) => action.Name === 'Add' || action.Name === 'Upload')
        setAvailable(isActionFound && !activeAction)
      } catch (error) {
        logger.error({
          message: localization.errorGettingActions,
          data: {
            details: { error },
          },
        })
      }
    }

    if (currentPath) {
      getActions()
    } else {
      setAvailable(false)
    }
  }, [localization.errorGettingActions, logger, repo, currentPath, activeAction])

  useEffect(() => {
    const getAllowedChildTypes = async () => {
      try {
        const allowedChildTypesFromRepo = await repo.allowedChildTypes.get({
          idOrPath: currentPath,
        })

        const filteredTypes = allowedChildTypesFromRepo.d.results
          .filter((type) => repo.schemas.getSchemaByName(type.Name).ContentTypeName === type.Name)
          .map((type) => repo.schemas.getSchemaByName(type.Name))

        const tempHasUpload = filteredTypes.some((type) => personalSettings.uploadHandlers.includes(type.HandlerName))

        setAllowedChildTypes(filteredTypes)
        setHasUpload(tempHasUpload)
      } catch (error) {
        logger.error({
          message: localization.errorGettingAllowedContentTypes,
          data: {
            details: { error },
          },
        })
      }
    }

    if (showSelectType) {
      currentPath ? getAllowedChildTypes() : setAllowedChildTypes([])
    }
  }, [
    localization.errorGettingAllowedContentTypes,
    logger,
    personalSettings.uploadHandlers,
    repo,
    showSelectType,
    currentPath,
  ])

  return (
    <div className={clsx(globalClasses.centered, globalClasses.relative)}>
      {!props.isOpened ? (
        <div className={globalClasses.drawerIconButtonWrapper}>
          {isAvailable ? (
            <Tooltip title={localization.addNew} placement="right">
              <span>
                <IconButton
                  className={globalClasses.drawerButton}
                  onClick={(event: React.MouseEvent<HTMLElement, MouseEvent>) => {
                    setAnchorEl(event.currentTarget)
                    setShowSelectType(true)
                  }}>
                  <Add className={globalClasses.drawerButtonIcon} />
                </IconButton>
              </span>
            </Tooltip>
          ) : (
            <span>
              <IconButton
                className={clsx(globalClasses.drawerButton, {
                  [classes.addButtonDisabled]: !isAvailable,
                })}
                onClick={(event: React.MouseEvent<HTMLElement, MouseEvent>) => {
                  setAnchorEl(event.currentTarget)
                  setShowSelectType(true)
                }}
                disabled={true}>
                <Add className={globalClasses.drawerButtonIcon} />
              </IconButton>
            </span>
          )}
        </div>
      ) : (
        <ListItem
          className={classes.listItem}
          button={true}
          onClick={(event: React.MouseEvent<HTMLElement, MouseEvent>) => {
            setAnchorEl(event.currentTarget)
            setShowSelectType(true)
          }}
          disabled={!isAvailable}>
          <ListItemIcon className={globalClasses.centeredHorizontal}>
            <Tooltip title={localization.addNew} placement="right">
              <span>
                <IconButton
                  className={clsx(globalClasses.drawerButtonExpanded, {
                    [classes.addButtonDisabled]: !isAvailable,
                  })}
                  disabled={!isAvailable}>
                  <Add className={globalClasses.drawerButtonIcon} />
                </IconButton>
              </span>
            </Tooltip>
          </ListItemIcon>
          <ListItemText primary={localization.addNew} />
        </ListItem>
      )}
      <Popover
        open={showSelectType}
        anchorEl={anchorEl}
        onClose={() => {
          setAnchorEl(null)
          setShowSelectType(false)
        }}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'left',
        }}>
        <List className={classes.listDropdown}>
          {hasUpload ? (
            <Tooltip title={localization.upload} placement="right">
              <ListItem
                key="Upload"
                button={true}
                style={{ padding: '10px 0 10px 10px' }}
                onClick={() => {
                  setShowSelectType(false)
                  openDialog({
                    name: 'upload',
                    props: { uploadPath: currentPath },
                    dialogProps: { open: true, fullScreen: false },
                  })
                }}>
                <ListItemIcon style={{ minWidth: '36px' }}>
                  <CloudUploadOutlined />
                </ListItemIcon>
                <ListItemText primary={localization.upload} className={classes.listItemTextDropdown} />
              </ListItem>
            </Tooltip>
          ) : null}

          {allowedChildTypes.length === 0 && !hasUpload ? (
            <div>{localization.noItems}</div>
          ) : (
            allowedChildTypes.map((childType) => (
              <Tooltip key={childType.ContentTypeName} title={childType.DisplayName} placement="right">
                <ListItem
                  button={true}
                  style={{ padding: '10px 0 10px 10px' }}
                  onClick={async () => {
                    setShowSelectType(false)
                    navigateToAction({
                      history,
                      routeMatch: snRoute.match,
                      action: 'new',
                      queryParams: { 'content-type': childType.ContentTypeName },
                    })
                  }}>
                  <ListItemIcon style={{ minWidth: '36px' }}>
                    <Icon item={childType} />
                  </ListItemIcon>
                  <ListItemText primary={childType.DisplayName} className={classes.listItemTextDropdown} />
                </ListItem>
              </Tooltip>
            ))
          )}
        </List>
      </Popover>
    </div>
  )
}
