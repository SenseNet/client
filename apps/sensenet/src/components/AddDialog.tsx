import Dialog, { DialogProps } from '@material-ui/core/Dialog/Dialog'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import { NewView } from '@sensenet/controls-react'
import { GenericContent, Schema } from '@sensenet/default-content-types'
import React, { useContext } from 'react'
import { LocalizationContext, RepositoryContext } from '../context'
import { LoggerContext } from '../context/LoggerContext'

export interface AddDialogProps {
  dialogProps: DialogProps
  schema: Schema
  parent: GenericContent
}

export const AddDialog: React.FunctionComponent<AddDialogProps> = ({ dialogProps, schema, parent }) => {
  const localization = useContext(LocalizationContext).values.addButton
  const repo = useContext(RepositoryContext)

  const handleClose = () => dialogProps.onClose && dialogProps.onClose(null as any)
  const logger = useContext(LoggerContext).withScope('AddDialog')

  return (
    <Dialog {...dialogProps}>
      <DialogTitle> {localization.dialogTitle.replace('{0}', schema.DisplayName)}</DialogTitle>
      <DialogContent>
        <NewView
          handleCancel={handleClose}
          repository={repo}
          contentTypeName={schema.ContentTypeName}
          schema={schema}
          path={parent.Path}
          title=""
          onSubmit={async (parentPath, content) => {
            try {
              const created = await repo.post({
                contentType: schema.ContentTypeName,
                parentPath,
                content,
              })
              handleClose()
              logger.information({
                message: localization.contentCreatedNotification.replace(
                  '{0}',
                  created.d.DisplayName || created.d.Name,
                ),
                data: {
                  relatedContent: created,
                  relatedRepository: repo.configuration.repositoryUrl,
                },
              })
            } catch (error) {
              logger.error({
                message: localization.errorGettingAllowedContentTypes,
                data: {
                  details: { error },
                },
              })
            }
          }}
        />
      </DialogContent>
    </Dialog>
  )
}
