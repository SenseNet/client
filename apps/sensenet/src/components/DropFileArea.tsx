import CloudUploadTwoTone from '@material-ui/icons/CloudUploadTwoTone'
import { GenericContent } from '@sensenet/default-content-types'
import React, { useState, useEffect } from 'react'
import { UploadProgressInfo } from '@sensenet/client-core'
import { ObservableValue } from '@sensenet/client-utils'
import { useInjector, useRepository, useTheme } from '../hooks'
import { UploadTracker } from '../services/UploadTracker'

export const DropFileArea: React.FunctionComponent<{
  parentContent: GenericContent
  style?: React.CSSProperties
}> = props => {
  const [isDragOver, setDragOver] = useState(false)

  const injector = useInjector()
  const repo = useRepository()
  const theme = useTheme()
  const [progressObservable] = useState(new ObservableValue<UploadProgressInfo>())

  useEffect(() => {
    const subscription = progressObservable.subscribe(p =>
      injector.getInstance(UploadTracker).onUploadProgress.setValue({ progress: p, repo }),
    )
    return () => subscription.dispose()
  }, [injector, progressObservable, repo])

  return (
    <div
      style={{
        position: 'relative',
        filter: isDragOver ? 'blur(1px)' : undefined,
        opacity: isDragOver ? 0.8 : 1,
        transition:
          'opacity 300ms cubic-bezier(0.445, 0.050, 0.550, 0.950), filter 300ms cubic-bezier(0.445, 0.050, 0.550, 0.950)',
        ...props.style,
      }}
      onDragEnter={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        setDragOver(true)
      }}
      onDragLeave={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        setDragOver(false)
      }}
      onDragOver={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        setDragOver(true)
      }}
      onDrop={ev => {
        ev.stopPropagation()
        ev.preventDefault()
        setDragOver(false)
        repo.upload.fromDropEvent({
          binaryPropertyName: 'Binary',
          createFolders: true,
          event: new DragEvent('drop', { dataTransfer: ev.dataTransfer }),
          overwrite: false,
          parentPath: props.parentContent ? props.parentContent.Path : '',
          progressObservable,
        })
      }}>
      <div
        style={{
          width: '100%',
          height: '100%',
          backgroundColor: theme.palette.primary.main,
          opacity: isDragOver ? 0.1 : 0,
          position: 'absolute',
          zIndex: -1,
          transition: 'opacity 300ms cubic-bezier(0.445, 0.050, 0.550, 0.950)',
        }}>
        <CloudUploadTwoTone style={{ width: '100%', height: '100%' }} />
      </div>
      {props.children}
    </div>
  )
}
