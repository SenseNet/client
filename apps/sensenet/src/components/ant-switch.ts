import { Switch } from '@material-ui/core'
import { createStyles, Theme, withStyles } from '@material-ui/core/styles'

export const AntSwitch = withStyles((theme: Theme) =>
  createStyles({
    root: {
      width: 28,
      height: 16,
      padding: 0,
      display: 'flex',
    },
    switchBase: {
      padding: 2,
      color: theme.palette.common.white,
      opacity: '1',
      '&$checked': {
        transform: 'translateX(12px)',
        color: theme.palette.common.black,
        '& + $track': {
          opacity: 1,
          backgroundColor: theme.palette.common.white,
        },
      },
    },
    thumb: {
      width: 12,
      height: 12,
      boxShadow: 'none',
    },
    track: {
      opacity: 1,
    },
    checked: {},
  }),
)(Switch)
