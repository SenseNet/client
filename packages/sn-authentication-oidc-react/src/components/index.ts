export * from './authenticating'
export * from './authentication-provider'
export * from './callback'
export * from './not-authenticated'
export * from './not-authorized'
export * from './oidc-routes'
export * from './oidc-secure'
export * from './session-lost'
export * from './silent-callback'
