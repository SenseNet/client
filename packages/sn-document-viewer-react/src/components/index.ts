export * from './comment'
export * from './document-widgets'
export * from './page-widgets'
export * from './ConfirmationDialog'
export * from './DocumentViewer'
export * from './DocumentViewerError'
export * from './DocumentViewerLayout'
export * from './DocumentViewerLoading'
export * from './DocumentViewerRegeneratePreviews'
export * from './LayoutAppBar'
export * from './Page'
export * from './PageList'
export * from './styles'
