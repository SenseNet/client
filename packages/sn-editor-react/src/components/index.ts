export * from './editor'
export * from './menu-bar'
