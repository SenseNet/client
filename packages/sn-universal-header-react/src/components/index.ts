export * from './hamburger-menu'
export * from './header'
export * from './three-dot-menu'
