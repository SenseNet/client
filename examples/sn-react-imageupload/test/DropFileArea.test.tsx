import { RepositoryContext } from '@sensenet/hooks-react'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { DropFileArea } from '../src/components/DropFileArea'

Object.defineProperty(window, 'DragEvent', {
  value: class DragEvent {},
})

describe('DropFileArea', () => {
  const testprop = {
    uploadsetdata: jest.fn(),
    notificationControll: jest.fn(),
    uploadPath: 'string',
    style: {
      display: 'flex',
      flexWrap: 'wrap',
    },
    setDragOver: jest.fn(),
    isDragOver: false,
  }
  it('Matches snapshot', () => {
    const wrapper = shallow(<DropFileArea {...testprop} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should set setDragOver on DragEnter', () => {
    const wrapper = mount(<DropFileArea {...testprop} />)
    act(() => {
      wrapper.find('.dropfilearea').prop('onDragEnter')!({
        stopPropagation: () => undefined,
        preventDefault: () => undefined,
      } as any)
    })

    expect(testprop.setDragOver).toBeCalledWith(true)
  })
  it('should set setDragOver on DragLeave', () => {
    const wrapper = mount(<DropFileArea {...testprop} />)
    act(() => {
      wrapper.find('.dropfilearea').prop('onDragLeave')!({
        stopPropagation: () => undefined,
        preventDefault: () => undefined,
      } as any)
    })

    expect(testprop.setDragOver).toBeCalledWith(false)
  })
  it('should set setDragOver on DragOver', () => {
    const wrapper = mount(<DropFileArea {...testprop} />)
    act(() => {
      wrapper.find('.dropfilearea').prop('onDragOver')!({
        stopPropagation: () => undefined,
        preventDefault: () => undefined,
      } as any)
    })

    expect(testprop.setDragOver).toBeCalledWith(true)
  })
  it('should set setDragOver and start Upload on Drop', async () => {
    const repository = {
      upload: {
        fromDropEvent: jest.fn(),
      },
    }

    const wrapper = mount(
      <RepositoryContext.Provider value={repository as any}>
        <DropFileArea {...testprop} />
      </RepositoryContext.Provider>,
    )
    await act(async () => {
      wrapper.find('.dropfilearea').prop('onDrop')!({
        stopPropagation: () => undefined,
        preventDefault: () => undefined,
      } as any)
    })

    expect(testprop.setDragOver).toBeCalledWith(false)
    expect(repository.upload.fromDropEvent).toBeCalled()
    expect(testprop.uploadsetdata).toBeCalled()
    expect(testprop.notificationControll).toBeCalled()
  })
})
