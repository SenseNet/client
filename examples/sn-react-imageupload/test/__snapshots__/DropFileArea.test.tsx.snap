// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropFileArea Matches snapshot 1`] = `
<div
  className="dropfilearea"
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  style={
    Object {
      "display": "inline-block",
      "filter": undefined,
      "flexWrap": "wrap",
      "opacity": 1,
      "transition": "opacity 300ms",
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "transparent",
        "height": "auto",
        "opacity": 0,
        "position": "relative",
        "transition": "opacity 300ms",
        "width": "auto",
      }
    }
  >
    <Memo([object Object])
      style={
        Object {
          "height": "250px",
          "marginTop": "0px",
          "position": "absolute",
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;
